core/ir.ml

```diff
@@ -39,103 +39,101 @@ type location = CommonTypes.Location.t
   [@@deriving show]
 
 type value =
-  [ `Constant of Constant.t
-  | `Variable of var
-  | `Extend of (value name_map * value option)
-  | `Project of (name * value)
-  | `Erase of (name_set * value)
-  | `Inject of (name * value * Types.datatype)
+  | Constant of Constant.t
+  | Variable of var
+  | Extend of (value name_map * value option)
+  | Project of (name * value)
+  | Erase of (name_set * value)
+  | Inject of (name * value * Types.datatype)
 
-  | `TAbs of tyvar list * value
-  | `TApp of value * tyarg list
+  | TAbs of (tyvar list * value)
+  | TApp of (value * tyarg list)
 
-  | `XmlNode of (name * value name_map * value list)
-  | `ApplyPure of (value * value list)
+  | XmlNode of (name * value name_map * value list)
+  | ApplyPure of (value * value list)
 
-  | `Closure of var * tyarg list * value
+  | Closure of (var * tyarg list * value)
 
-  | `Coerce of (value * Types.datatype)
-  ]
+  | Coerce of (value * Types.datatype)
 and tail_computation =
-  [ `Return of (value)
-  | `Apply of (value * value list)
-  (* | `ApplyClosure of (value * value list) *)
+  | Return of (value)
+  | Apply of (value * value list)
+  (* | ApplyClosure of (value * value list) *)
 
-  | `Special of special
+  | Special of special
 
-  | `Case of (value * (binder * computation) name_map * (binder * computation) option)
-  | `If of (value * computation * computation)
-  ]
-and fun_def = binder * (tyvar list * binder list * computation) * binder option * location
+  | Case of (value * (binder * computation) name_map * (binder * computation) option)
+  | If of (value * computation * computation)
+and fun_def = (binder * (tyvar list * binder list * computation) * binder option * location)
 and binding =
-  [ `Let of binder * (tyvar list * tail_computation)
-  | `Fun of fun_def
-  | `Rec of fun_def list
-  | `Alien of (binder * name * language)
-  | `Module of (string * binding list option) ]
+  | Let of (binder * (tyvar list * tail_computation))
```

Comment:
I think we should get rid of the needless boxing before merging this, e.g. drop the outermost parentheses (with regular variants they have performance implications).

Reply:
Or if you solemnly swear to remove the parentheses in a follow-up PR, then I will not object to merging this now.

Reply:
done!

Reply:
Why are the outermost parentheses a problem? I'm asking cause I introduced a fair amount of this in my refactoring. The reason was that with polymorphic variants I had something like:

```OCaml
type foo = [ `A of int * int ]

match f with 
  | `A a -> ()
```
After converting `foo` to an ordinary variant the pattern match `A a` was no longer valid, with compiler reporting an error that `A` expects two arguments but was only give one. My solution was to introduce parentheses around the arguments in the datatype definition:

```OCaml
type foo = A of (int * int)
```

Reply:
> Why are the outermost parentheses a problem?

They introduce an extra box. For regular variants the code produced by following snippets differ
```
type t = Foo of int * int | Bar
let _ = Foo (2,3)
```
and
```ocaml
type t = Foo of (int * int) | Bar
let _ = Foo (2,3)
```
The former is represented `[0: 2 3]` at runtime, whilst the latter is represented as `[0: [0: 2 3]]`, that is there is an extra box. For polymorphic variants this doesn't matter because the constructor arguments are always boxed.

Reply:
What if I said `type t = Foo of ((((((int * int)))))) | Bar`? Would it mean six additional layers of indirection?

Reply:
> type t = Foo of ((((((int * int)))))) | Bar

No, only the first pair of parentheses have a semantic effect in this example.

Reply:
However, 
```ocaml
type t = Foo of ((((((int * int)))) * bool)) | Bar
let _ = Foo ((2,3), true)
```
would be represented as `[0: [0: 2 3] 1a]` at runtime.

Reply:
Good to know. I was about to ask about all of this on the last dev meeting but we ran out of time.

I'll remove these extra parentheses in Sugartypes in near future.